io/io/test/rfile.cxx

```diff
@@ -0,0 +1,500 @@
+#include "gtest/gtest.h"
+#include "gmock/gmock.h"
+
+#include <TFile.h>
+#include <TH1D.h>
+#include <TROOT.h>
+#include <TSystem.h>
+#include <ROOT/RError.hxx>
+#include <ROOT/RFile.hxx>
+#include <ROOT/TestSupport.hxx>
+#include <numeric>
+
+using ROOT::Experimental::RFile;
+
+namespace {
+
+/**
+ * An RAII wrapper around an open temporary file on disk. It cleans up the guarded file when the wrapper object
+ * goes out of scope.
+ */
+class FileRaii {
+private:
+   std::string fPath;
+   bool fPreserveFile = false;
+
+public:
+   explicit FileRaii(const std::string &path) : fPath(path) {}
+   FileRaii(FileRaii &&) = default;
+   FileRaii(const FileRaii &) = delete;
+   FileRaii &operator=(FileRaii &&) = default;
+   FileRaii &operator=(const FileRaii &) = delete;
+   ~FileRaii()
+   {
+      if (!fPreserveFile)
+         std::remove(fPath.c_str());
+   }
+   std::string GetPath() const { return fPath; }
+
+   // Useful if you want to keep a test file after the test has finished running
+   // for debugging purposes. Should only be used locally and never pushed.
+   void PreserveFile() { fPreserveFile = true; }
+};
+
+} // anonymous namespace
+
+TEST(RFile, Open)
+{
+   FileRaii fileGuard("test_rfile_read.root");
+
+   // Create a root file to open
+   {
+      auto tfile = std::unique_ptr<TFile>(TFile::Open(fileGuard.GetPath().c_str(), "RECREATE"));
+      TH1D hist("hist", "", 100, -10, 10);
+      hist.FillRandom("gaus", 1000);
+      tfile->WriteObject(&hist, "hist");
+   }
+
+   auto file = RFile::Open(fileGuard.GetPath());
+   auto hist = file->Get<TH1D>("hist");
+   EXPECT_TRUE(hist);
+
+   EXPECT_FALSE(file->Get<TH1D>("inexistent"));
+   EXPECT_FALSE(file->Get<TH1F>("hist"));
+   EXPECT_TRUE(file->Get<TH1>("hist"));
+
+   // We do NOT want to globally register RFiles ever.
+   EXPECT_EQ(ROOT::GetROOT()->GetListOfFiles()->GetSize(), 0);
+
+   std::string foo = "foo";
+   EXPECT_THROW(file->Put("foo", foo), ROOT::RException);
+}
+
+TEST(RFile, OpenInexistent)
+{
+   FileRaii fileGuard("does_not_exist.root");
+
+   // make sure that the file really does not exist, in case a previous test didn't clean it up.
```

Comment:
Is there any test that intentionally create the file named 'does_not_exist.root' ?  The comment seems to imply there is.